### ISSUE TITLE
Pinning transitive dependencies with a lockfile

### ISSUE THREAD

**Issue by delan** [platformio/platformio-core#4613]
- [x] **Feature Request**.

------------------------------------------------------------------

### Configuration

**Operating system**: Windows 10

**PlatformIO Version** (`platformio --version`): 6.1.6

### Description of problem

It’s good practice to [pin your dependencies to exact versions](https://docs.platformio.org/en/latest/librarymanager/dependencies.html#declaring-practices) to avoid unexpected regressions, but transitive dependency updates can also cause unexpected regressions, and pinning your transitive dependencies is currently painful.

You can check in your project’s `.pio` directory to version control, which has the added benefit of protecting against packages being changed or removed, but it only includes your lib_deps and not your platform_packages. You can add everything in your `pio pkg list` tree to your platform_packages and lib_deps, but you have to be careful to put the deepest dependencies first (platformio/platformio-core#2115), and mixing direct dependencies with transitive dependencies makes updating hard.

Many libraries don’t constrain their dependencies at all. Even if all the libraries you use constrain their own dependencies to those that are [semver compatible](https://docs.platformio.org/en/latest/core/userguide/pkg/cmd_install.html#version-requirements), the only way to avoid breakage entirely (and have [reproducible builds](https://reproducible-builds.org)) without pinning your transitive dependencies is to only depend on libraries that pin everything to exact versions, which is impossible in practice.

Having a lockfile like Rust’s [Cargo.lock](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html) or npm’s [package-lock.json](https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json) that lists all of your direct and transitive dependencies, with their exact version numbers *and* hashes of their contents (in case they are modified after release), would prevent most\* of these problems. When installing packages for a project, we would use the lockfile in the top-level project if any, and ignore any lockfiles in dependencies.

\* it won’t prevent packages being removed, but the only way to do that without a centralised package repository is to vendor your entire dependency tree (copy everything into your project)

**Comment by winksaville** [comment]
In my personal opinion the lack of this feature makes PlatformIO a non-starter. I've subscribed to this issue and will be back when there is at least a prototype of an automated implemention and a commitment to provide this feature.

**Comment by m-mcgowan** [comment]
I agree, it's been frustrating to try to create reproducible builds.  I've coded (and to be honest, only lightly tested) an extension that attempts to pin all the components of a platformio build for reproducible builds.  My primary focus has been on the pioarduino espressif32 platform, but I've tried to avoid anything platform-specific in the implementation.

You can find it here: https://github.com/m-mcgowan/pio-lock.

It also avoids the non-deterministic output that you normally get by setting the source time so that `__DATE__` and `__TIME__` expand to deterministic values - and reuse those values when rebuilding a previous release.
